lib/stripe/credit_card_cloner.rb

```diff
@@ -1,38 +1,115 @@
+        next unless customer_id
+
+        customer = Stripe::Customer.retrieve(customer_id, stripe_account: stripe_account)
+        customer&.delete unless customer.deleted?
+      end
+    end
+
+    private
+
     def clone(credit_card, connected_account_id)
       new_payment_method = clone_payment_method(credit_card, connected_account_id)
 
       # If no customer is given, it will clone the payment method only
-      return nil, new_payment_method.id if credit_card.gateway_customer_profile_id.blank?
+      return [nil, new_payment_method.id] if credit_card.gateway_customer_profile_id.blank?
 
       new_customer = Stripe::Customer.create({ email: credit_card.user.email },
                                              stripe_account: connected_account_id)
       attach_payment_method_to_customer(new_payment_method.id,
                                         new_customer.id,
                                         connected_account_id)
 
+      add_metadata_to_payment_method(new_payment_method.id, connected_account_id)
+
       [new_customer.id, new_payment_method.id]
     end
 
-    private
+    def find_cloned_card(card, connected_account_id)
+      return nil unless fingerprint = fingerprint_for_card(card)
+
+      find_customers(card.user.email, connected_account_id).each do |customer|
+        find_payment_methods(customer.id, connected_account_id).each do |payment_method|
+          if payment_method_is_clone?(payment_method, fingerprint)
+            return [customer.id, payment_method.id]
+          end
+        end
+      end
+      nil
+    end
+
+    def payment_method_is_clone?(payment_method, fingerprint)
+      payment_method.card.fingerprint == fingerprint && payment_method.metadata["ofn-clone"]
+    end
+
+    def fingerprint_for_card(card)
+      Stripe::PaymentMethod.retrieve(card.gateway_payment_profile_id).card.fingerprint
+    end
+
+    def find_customers(email, connected_account_id)
+      start_after, customers = nil, []
+
+      (1..request_limit = 100).each do |request_number|
+        response = Stripe::Customer.list({ email: email, starting_after: start_after, limit: 100 },
+                                         stripe_account: connected_account_id)
+        customers += response.data
+        break unless response.has_more
+
+        start_after = response.data.last.id
+        notify_limit(request_number, "customers") if request_limit == request_number
+      end
+      customers
+    end
+
+    def find_payment_methods(customer_id, connected_account_id)
+      start_after, payment_methods = nil, []
+
+      (1..request_limit = 10).each do |request_number|
+        options = { customer: customer_id, type: 'card', starting_after: start_after, limit: 100 }
+        response = Stripe::PaymentMethod.list(options, stripe_account: connected_account_id)
+        payment_methods += response.data
+        break unless response.has_more
```

Comment:
Stripe's Ruby client already deals with pagination itself :point_right: https://stripe.com/docs/api/pagination/auto?lang=ruby. This means we can directly use `#auto_paging_each` at line 63 instead of this whole helper method.

Reply:
🤯

Reply:
13b95f41bbb1e6366decc8282829febe1e7b9446